README.md
hamstr: Hierarchical Accumulation Modelling with Stan and R.
================
Andrew M. Dolman
2021-06-17

------------------------------------------------------------------------

**hamstr** implements a *Bacon-like* (Blaauw and Christen, 2011)
sediment accumulation or age-depth model with hierarchically structured
multi-resolution sediment sections. The Bayesian model is implemented in
the Stan probabilistic programming language (<https://mc-stan.org/>).

## Installation

**hamstr** can be installed directly from Github

``` r
if (!require("remotes")) {
  install.packages("remotes")
}

remotes::install_github("earthsystemdiagnostics/hamstr", args = "--preclean", build_vignettes = FALSE)
```

## Using **hamstr**

Examples using the example core “MSB2K” from the
[rbacon](https://cran.r-project.org/web/packages/rbacon/index.html)
package.

``` r
library(hamstr)
library(rstan)
library(tidyverse)
#> Warning: package 'tibble' was built under R version 4.0.3
#> Warning: package 'readr' was built under R version 4.0.3

set.seed(20200827)
```

### Converting radiocarbon ages to calendar ages.

Unlike Bacon, **hamstr** does not do the conversion of radiocarbon dates
to calendar ages as part of the model fitting process. This must be done
in advance. **hamstr** includes the helper function `calibrate_14C_age`
to do this, which in turn uses the function `BchronCalibrate` from the
[Bchron](https://cran.r-project.org/web/packages/Bchron/index.html)
package.

Additionally, unlike Bacon, **hamstr** approximates the complex
empirical calendar age PDF that results from calibration into a single
point estimate and 1-sigma uncertainty. This is a necessary compromise
in order to be able to use the power of the Stan platform. Viewed in
context with the many other uncertainties in radiocarbon dates and the
resulting age-models this will not usually be a major issue.

The function `calibrate_14C_age` will append columns to a data.frame
with the calendar ages and 1-sigma uncertainties.

``` r
MSB2K_cal <- calibrate_14C_age(MSB2K, age.14C = "age", age.14C.se = "error")
```

The approximated calendar age PDFs can be compared with the empirical
PDFs with the function `compare_14C_PDF`

A sample of six dates are plotted here for the IntCal20 and Marine20
calibrations. This approximation is much less of an issue for marine
radiocarbon dates, as the cosmogenic radiocarbon signal has been
smoothed by mixing in the ocean.

``` r
i <- seq(1, 40, by = floor(40/6))[1:6]
compare_14C_PDF(MSB2K$age[i], MSB2K$error[i], cal_curve = "intcal20")+
  labs(title = "Intcal20")
```

![](readme_files/figure-gfm/unnamed-chunk-4-1.png)<!-- -->

``` r
compare_14C_PDF(MSB2K$age[i], MSB2K$error[i], cal_curve = "marine20") +
  labs(title = "Marine20")
```

![](readme_files/figure-gfm/unnamed-chunk-5-1.png)<!-- -->

### Fitting age-models with **hamstr**

By default **hamstr** runs with three Markov chains and these can be run
in parallel. This code will assign 3 (processor) cores as long as the
machine has at least 3. The number of cores can also be set for specific
calls of the `hamstr` function using the `cores` argument.

``` r
if (parallel::detectCores() >= 3) options(mc.cores = 3)
```

Age-depth (sediment accumulation) models are fit with the function
`hamstr`. A vectors of depth, observed age and age uncertainty are
passed as arguments to the function.

``` r
hamstr_fit_1 <- hamstr(depth = MSB2K_cal$depth,
                       obs_age = MSB2K_cal$age.14C.cal,
                       obs_err = MSB2K_cal$age.14C.cal.se)
```

The default plotting method shows the fitted age models together with
some diagnostic plots: a traceplot of the log-posterior to assess
convergence of the overall model; a plot of accumulation rate against
depth at each hierarchical level; the prior and posterior of the memory
parameter. By default the age-models are summarised to show the mean,
median, 25% and 95% posterior intervals. The data are shown as points
with their 1-sigma uncertainties. The structure of the sections is shown
along the top of the age-model plot.

``` r
plot(hamstr_fit_1)
#> Joining, by = "idx"
#> Joining, by = "alpha_idx"
```

![](readme_files/figure-gfm/unnamed-chunk-8-1.png)<!-- -->

A “spaghetti” plot can be created instead of shaded regions. This shows
a random sample of iterations from the posterior distribution
(realisation of the age-depth model). This can be slow if lots of
iterations are plotted, the default is to plot 1000 iterations.
Additionally, plotting of the diagnostic plots can be switched off.

``` r
plot(hamstr_fit_1, summarise = FALSE, plot_diagnostics = FALSE)
#> Joining, by = "idx"
```

![](readme_files/figure-gfm/unnamed-chunk-9-1.png)<!-- -->

#### Mean accumulation rate

There is no need to specify a prior value for the mean accumulation
rate, parameter `acc.mean` in Bacon, as in **hamstr**, this overall mean
accumulation rate is a full parameter estimated from the data.

By default, **hamstr** uses robust linear regression (`MASS::rlm`) to
estimate the mean accumulation rate from the data, and then uses this to
parametrise a prior distribution for the overall mean accumulation rate.
This prior is a half-normal with zero mean and standard deviation equal
to 10 times the estimated mean. Although this does introduce an element
of “data-peaking”, using the data twice (for both the prior and
likelihood), the resulting prior is only weakly-informative. The
advantage is that the prior is automatically scaled appropriately
regardless of the units of depth or age.

This prior can be checked visually against the posterior.

``` r
plot(hamstr_fit_1, type = "acc_mean_prior_post")
```

![](readme_files/figure-gfm/unnamed-chunk-10-1.png)<!-- -->

Default parameter values for the shape of the accumulation prior
`acc_shape = 1.5`, the memory mean `mem_mean = 0.5` and strength
`mem_strength = 10`, are the same as for Bacon &gt;= 2.5.1.

#### Setting the thickness and number of discrete sections

One of the more critical tuning parameters in the **Bacon** model is the
parameter `thick`, which determines the thickness and number of discrete
core sections. Finding a good or optimal value for a given core is often
critical to getting a good age-depth model. Too few sections and the
resulting age-model is very “blocky” and can miss changes in
sedimentation rate; however, counter-intuitively, too many very thin
sections can also often result in an age-model that “under-fits” the
data - ploughing a straight line through the age-control points when a
lower resolution model shows variation in accumulation rate.

The key structural difference between **Bacon** and **hamstr** models is
that with **hamstr** the sediment core is modelled at multiple
resolutions simultaneously with a hierarchical structure. This removes
the need to trade-off smoothness and flexibility.

The parameter `K` controls the number and structure of the hierarchical
sections. It is specified as a vector, where each value indicates the
number of new child sections for each parent section at each finer
hierarchical level. E.g. `c(10, 10)` would specify 10 sections at the
coarsest level, with 10 new sections at the next finer level for each
coarse section, giving a total of 100 sections at the highest / finest
resolution level. `c(10, 10, 10)` would specify 1000 sections at the
finest level and 3 hierarchical levels of 10, 100 and 1000 sections.

The structure is hierarchical in the sense that the modelled
accumulation rates for the parent sections act as priors for their child
sections; specifically, the mean accumulation rate for a given parent is
the mean of the gamma prior for it’s child sections. In turn, the
overall mean accumulation rate for the whole core is itself a parameter
estimated by the fitting process. The hierarchical structure of
increasing resolution allows the model to adapt to low-frequency changes
in the accumulation rate, that is changes between “regimes” of high or
low accumulation that persist for long periods.

By default `K` is chosen so that the number of hierarchical levels, and
the number of new child sections per level, are approximately equal,
e.g. c(4, 4, 4, 4). The total number of sections at the finest level is
set so that the resolution is 1 cm per section, up to a total length of
900 cm, above which the default remains 900 sections and a coarser
resolution is used. This can changed with the parameter K.

Here is an example of a coarser model.

``` r
hamstr_fit_2 <- hamstr(depth = MSB2K_cal$depth,
                   obs_age = MSB2K_cal$age.14C.cal,
                   obs_err = MSB2K_cal$age.14C.cal.se,
                   K = c(5, 5))
#> 
#> SAMPLING FOR MODEL 'hamstr' NOW (CHAIN 1).
#> Chain 1: 
#> Chain 1: Gradient evaluation took 0 seconds
#> Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 0 seconds.
#> Chain 1: Adjust your expectations accordingly!
#> Chain 1: 
#> Chain 1: 
#> Chain 1: Iteration:    1 / 2000 [  0%]  (Warmup)
#> Chain 1: Iteration:  200 / 2000 [ 10%]  (Warmup)
#> Chain 1: Iteration:  400 / 2000 [ 20%]  (Warmup)
#> Chain 1: Iteration:  600 / 2000 [ 30%]  (Warmup)
#> Chain 1: Iteration:  800 / 2000 [ 40%]  (Warmup)
#> Chain 1: Iteration: 1000 / 2000 [ 50%]  (Warmup)
#> Chain 1: Iteration: 1001 / 2000 [ 50%]  (Sampling)
#> Chain 1: Iteration: 1200 / 2000 [ 60%]  (Sampling)
#> Chain 1: Iteration: 1400 / 2000 [ 70%]  (Sampling)
#> Chain 1: Iteration: 1600 / 2000 [ 80%]  (Sampling)
#> Chain 1: Iteration: 1800 / 2000 [ 90%]  (Sampling)
#> Chain 1: Iteration: 2000 / 2000 [100%]  (Sampling)
#> Chain 1: 
#> Chain 1:  Elapsed Time: 2.354 seconds (Warm-up)
#> Chain 1:                1.965 seconds (Sampling)
#> Chain 1:                4.319 seconds (Total)
#> Chain 1: 
#> 
#> SAMPLING FOR MODEL 'hamstr' NOW (CHAIN 2).
#> Chain 2: 
#> Chain 2: Gradient evaluation took 0 seconds
#> Chain 2: 1000 transitions using 10 leapfrog steps per transition would take 0 seconds.
#> Chain 2: Adjust your expectations accordingly!
#> Chain 2: 
#> Chain 2: 
#> Chain 2: Iteration:    1 / 2000 [  0%]  (Warmup)
#> Chain 2: Iteration:  200 / 2000 [ 10%]  (Warmup)
#> Chain 2: Iteration:  400 / 2000 [ 20%]  (Warmup)
#> Chain 2: Iteration:  600 / 2000 [ 30%]  (Warmup)
#> Chain 2: Iteration:  800 / 2000 [ 40%]  (Warmup)
#> Chain 2: Iteration: 1000 / 2000 [ 50%]  (Warmup)
#> Chain 2: Iteration: 1001 / 2000 [ 50%]  (Sampling)
#> Chain 2: Iteration: 1200 / 2000 [ 60%]  (Sampling)
#> Chain 2: Iteration: 1400 / 2000 [ 70%]  (Sampling)
#> Chain 2: Iteration: 1600 / 2000 [ 80%]  (Sampling)
#> Chain 2: Iteration: 1800 / 2000 [ 90%]  (Sampling)
#> Chain 2: Iteration: 2000 / 2000 [100%]  (Sampling)
#> Chain 2: 
#> Chain 2:  Elapsed Time: 2.014 seconds (Warm-up)
#> Chain 2:                1.74 seconds (Sampling)
#> Chain 2:                3.754 seconds (Total)
#> Chain 2: 
#> 
#> SAMPLING FOR MODEL 'hamstr' NOW (CHAIN 3).
#> Chain 3: 
#> Chain 3: Gradient evaluation took 0 seconds
#> Chain 3: 1000 transitions using 10 leapfrog steps per transition would take 0 seconds.
#> Chain 3: Adjust your expectations accordingly!
#> Chain 3: 
#> Chain 3: 
#> Chain 3: Iteration:    1 / 2000 [  0%]  (Warmup)
#> Chain 3: Iteration:  200 / 2000 [ 10%]  (Warmup)
#> Chain 3: Iteration:  400 / 2000 [ 20%]  (Warmup)
#> Chain 3: Iteration:  600 / 2000 [ 30%]  (Warmup)
#> Chain 3: Iteration:  800 / 2000 [ 40%]  (Warmup)
#> Chain 3: Iteration: 1000 / 2000 [ 50%]  (Warmup)
#> Chain 3: Iteration: 1001 / 2000 [ 50%]  (Sampling)
#> Chain 3: Iteration: 1200 / 2000 [ 60%]  (Sampling)
#> Chain 3: Iteration: 1400 / 2000 [ 70%]  (Sampling)
#> Chain 3: Iteration: 1600 / 2000 [ 80%]  (Sampling)
#> Chain 3: Iteration: 1800 / 2000 [ 90%]  (Sampling)
#> Chain 3: Iteration: 2000 / 2000 [100%]  (Sampling)

NEWS.md
# hamstr 0.6.0

* Ability added to model age-heterogeneity from bioturbation and its effect on age models.
* Default plot changed to put diagnostic plots at the bottom.


# hamstr 0.5.2

* Minor change to the way acc_shape is adjusted to account for the number of hierarchical levels. acc_shape is now adjusted to control the shape of the gamma distributed alpha parameters, not the total variance. The difference is minor.

* New methods to extract and plot accumulation rates


## hamstr 0.5.1

* New default hierarchical structure K based on default resolution of 1 cm (up to 1000 cm)
and number of levels approximately equal to number of new child sections per parent per level.


## hamstr 0.5.0

* Significant update to behaviour when using hierarchical structure. The acc_shape 
parameter is now adjusted to control the total variance in the alpha parameters 
when there are multiple levels
* Default mem_mean and mem_strength values updated to 0.5 and 10 to match 
updated defaults from bacon 2.5.1 onwards


## hamstr 0.4.1

* Methods written for generic functions plot and summary


## hamstr 0.4

* Package and main function renamed to hamstr (hierarchical accumulation models with Stan and R)
* dev branch with hierarchical sections merged to master


## baconr 0.2

* Hiatuses can now be modelled
* Added more information to the summary figure


package-dev-scripts/compare-hamstr-bacon.R
#devtools::install(quick = FALSE, dependencies = FALSE)
# install.packages("../hamstr", repos = NULL, type = "source")
library(hamstr)
library(tidyverse)

MSB2K_cal <- hamstr::calibrate_14C_age(MSB2K, age.14C = "age", age.14C.se = "error")


###
# simulate correlated gamma?

SimulateAgeDepth <- function(top, bottom, d_depth, gamma_shape,
                             gamma_mean, gamma_breaks = NULL,
                             ar1){

  pars <- c(as.list(environment()))


  depth <- seq(top, bottom, by = d_depth)

  if (is.null(gamma_breaks) == FALSE){
    if(length(gamma_mean) - length(gamma_breaks) != 1)
      stop("gamma_mean should be 1 longer than gamma_breaks")

    foo <- stepfun(gamma_breaks, gamma_mean)

    gamma_mean <- foo(depth[-1])
  }


  n <- length(depth)

  gamma_scale = gamma_mean / gamma_shape

  y <- as.numeric(arima.sim(model = list(ar = ar1^d_depth), n-1))
  y <- y / sd(y)
  y <- y - mean(y)
  sd_y <- sd(y)

  py <- (pnorm(y, mean = 0, sd = sd_y))

  # translate to gamma
  acc.rates_yr_cm <- qgamma(py, shape = gamma_shape, scale = gamma_scale)

  min.age <- gamma_mean[1] * top

  age <- cumsum(c(min.age, acc.rates_yr_cm) * d_depth)


  core <- dplyr::tibble(depth = depth, age = age, acc.rates_yr_cm = c(acc.rates_yr_cm, NA))

  return(list(pars = pars, core = core))

  }



tmp <- SimulateAgeDepth(top = 100, bottom = 700, d_depth = 1.1,
                        gamma_shape = 1.5,
                        gamma_mean = c(50, 10, 50), gamma_breaks = c(200, 600),
                        ar1 = 0.7)

tmp$core %>%
  ggplot(aes(depth, age)) +
  geom_line()



SimulateCarbonDates <- function(age_depth_obj, sampling_interval, cal_curve,
                                sample_gap = NULL){

  pars <- c(as.list(environment()))

  obs <- age_depth_obj$core %>%
    mutate(sampled = (rank(depth) %% sampling_interval == 0)) %>%
    filter(sampled) %>%
    mutate(age_14C = Bchron::unCalibrate(age, type = "ages")) %>%
    mutate(age_14C_sigma = 20 + age * 0.02) %>%
    mutate(age_14C_hat = rnorm(n(), age_14C, age_14C_sigma))

  if (is.null(sample_gap) == FALSE){
    obs <- obs %>%
      filter(depth < sample_gap[1] | depth > sample_gap[2])
  }

  obs <- obs %>%
    select(depth, age_14C_hat, age_14C_sigma) %>%
    calibrate_14C_age(.,
                      "age_14C_hat", "age_14C_sigma",
                      cal_curve = cal_curve)


  return(list(obs = obs,
              core = age_depth_obj$core,
              pars = pars))

}

## should be comparing coverage of simulated "true" ages not the data points
## as data might have extra error added
join_mod_dat <- function(hamstr_fit, dat){

  #browser()
  i1 <- predict(hamstr_fit, dat$depth)
  s1 <- summary(i1)

  # try a bind rather than join, should be much faster.
  c1 <- dplyr::left_join(dat, s1)

  #c1 <- bind_cols(dat, s1)


  return(c1)
}


AgeModCoverage <- function(hamstr_fit, dat){
  comb <- join_mod_dat(hamstr_fit, dat)

  comb %>%
    filter(complete.cases(`75%`)) %>%
    mutate(c95 = (age >= `2.5%`) & (age <= `97.5%`),
           c50 = (age >= `25%`) & (age <= `75%`)) %>%
    #select(c95)
    summarise(c95 = sum(c95) / n(),
              c50 = sum(c50) / n(),
              RMSE = sqrt(mean((age - `50%`)^2)))

}


### Compare hamstr and Bacon

CompareHamBac <- function(age_depth_obj,
                          ar_coefs, acc_mean,
                          mem_mean, mem_strength, acc_shape,
                          scale_shape = TRUE,
                          K_hamstr, K_bacon,
                          inflate_errors = FALSE){



  ham1 <-  hamstr(depth = age_depth_obj$obs$depth, obs_age = age_depth_obj$obs$age.14C.cal,
                  obs_err = age_depth_obj$obs$age.14C.cal.se,
                  top_depth = min(age_depth_obj$core$depth),
                  bottom_depth = max(age_depth_obj$core$depth),
                  K = K_hamstr,
                  acc_shape = acc_shape,
                  mem_mean = mem_mean, mem_strength = mem_strength,
                  inflate_errors = inflate_errors,
                  scale_shape = scale_shape,
                  chains = 3, cores = 3)


  bac1 <- hamstr_bacon(depth = age_depth_obj$obs$depth,
                       obs_age = age_depth_obj$obs$age_14C_hat,
                       obs_err = age_depth_obj$obs$age_14C_sigma,
                       acc.mean = acc_mean, acc.shape = acc_shape,
                       mem.mean = mem_mean, mem.strength = mem_strength,
                       d.min = min(age_depth_obj$core$depth),
                       d.max = max(age_depth_obj$core$depth),
                       thick = diff(range(age_depth_obj$core$depth)) / K_bacon,
                       suggest = "FALSE")

  return(list(age_depth_obj = age_depth_obj,
              hamstr = ham1, bacon = bac1
  ))
}



GetCoverages <- function(sim){

  hamstr = AgeModCoverage(sim$hamstr, sim$age_depth_obj$core)
  bacon = AgeModCoverage(sim$bacon, sim$age_depth_obj$core)

  bind_rows(hamstr=hamstr, bacon=bacon, .id = "method")

}

PlotHamstrBacon <- function(sim){

  obs_ages <- hamstr::calibrate_14C_age(sim$bacon$data, age.14C = "age", age.14C.se = "error")

  obs_ages <- obs_ages %>%
    dplyr::select(-age, -error) %>%
    dplyr::rename(.,
                  age = age.14C.cal,
                  err  = age.14C.cal.se) %>%
    dplyr::mutate(age_upr = age + 2*err,
                  age_lwr = age - 2*err)

  bacon_age_summary <- hamstr:::summarise_bacon_age_models(sim$bacon)

  hamstr_age_summary <- hamstr:::summarise_age_models(sim$hamstr)

  alpha.lvl <- 0.5
  #pal <- c("Hamstr" = "Blue", "Bacon" = "Red", "Simulated" = "Yellow")
  pal <- c("Hamstr" = "#7570b3", "Bacon" = "#e7298a", "Simulated" = "green")

  p.age.sum <- bacon_age_summary %>%
    ggplot2::ggplot(ggplot2::aes(x = depth, y = mean)) +
    ggplot2::geom_ribbon(ggplot2::aes(ymax = `2.5%`, ymin = `97.5%`, fill = "Bacon"), alpha = alpha.lvl/1.5) +
    ggplot2::geom_ribbon(ggplot2::aes(ymax = `75%`, ymin = `25%`, fill = "Bacon"), alpha = alpha.lvl) +
    #ggplot2::geom_line() +
    ggplot2::geom_line(ggplot2::aes(y = `50%`, colour = "Bacon"), lwd = 1) +
    ggplot2::labs(x = "Depth", y = "Age") #+
  #ggplot2::theme(legend.key = element_rect(fill  = "grey40", colour = "grey40"))#+
  #ggplot2::theme_bw()# +
  #ggplot2::theme(panel.grid = ggplot2::element_blank()) +


  p.age.sum <- p.age.sum +
    ggplot2::geom_ribbon(data = hamstr_age_summary,
                         ggplot2::aes(ymax = `2.5%`, ymin = `97.5%`, fill = "Hamstr"), alpha = alpha.lvl/1.5) +
    ggplot2::geom_ribbon(data = hamstr_age_summary,
                         ggplot2::aes(ymax = `75%`, ymin = `25%`, fill = "Hamstr"), alpha = alpha.lvl) +
    #ggplot2::geom_line(data = hamstr_age_summary ) +
    ggplot2::geom_line(data = hamstr_age_summary, ggplot2::aes(y = `50%`, colour = "Hamstr"), lwd = 1)

  p.age.sum <- p.age.sum +
    ggplot2::geom_linerange(data = obs_ages,
                            ggplot2::aes(x = depth,
                                         ymax = age_upr, ymin = age_lwr), inherit.aes = FALSE,
                            colour = "green") +
    ggplot2::geom_point(data = obs_ages, ggplot2::aes(y = age),
                        colour = "green") +
    geom_line(data = sim$age_depth_obj$core, aes(x = depth, y = age, colour = "Simulated"), lwd = 1)+

    scale_fill_manual("", values = pal) +
    scale_color_manual("", values = pal)


  p.age.sum

}

CompareMedAgeMod <- function(sim){

  hamstr = join_mod_dat(sim$hamstr, sim$sim_core)
  bacon = join_mod_dat(sim$bacon, sim$sim_core)

  hb <- bind_rows(hamstr=hamstr, bacon=bacon, .id = "method") %>%
    mutate(error = `50%` - age)

  p <- hb %>%
    ggplot(aes(x = depth, y = `50%` - age, colour = method)) +
    geom_line() +
    geom_hline(yintercept = 0)

  return(list(data = hb, plot = p))

}




ad.obj <- SimulateAgeDepth(top = 100, bottom = 700, d_depth = 1.1,
                           gamma_shape = 1.5,
                           gamma_mean = c(50, 10, 50), gamma_breaks = c(200, 600),
                           ar1 = 0.5)

tmp <- SimulateCarbonDates(ad.obj, sampling_interval = 24, sample_gap = c(201, 349),
                    cal_curve = "intcal20")

tst1 <- CompareHamBac(tmp,
                       K_hamstr = c(10,10),
                       K_bacon = 100,
                      scale_shape = FALSE,
                       acc_shape = 1.5,
                      acc_mean = mean(tmp$pars$age_depth_obj$pars$gamma_mean),
                       mem_mean = 0.5, mem_strength = 10)

tst2 <- CompareHamBac(tmp,
                      K_hamstr = c(10,10),
                      K_bacon = 100,
                      scale_shape = TRUE,
                      acc_shape = 1.5,
                      acc_mean = mean(tmp$pars$age_depth_obj$pars$gamma_mean),
                      mem_mean = 0.5, mem_strength = 10)


PlotHamstrBacon(tst1)
GetCoverages(tst1)

PlotHamstrBacon(tst2)
GetCoverages(tst2)




## Simulations ------

# hbc1 <- CompareHamBac(100, 400, 1, gamma_shape = 1.5, gamma_mean = 20, ar_coefs = 0.7,
#               sampling_interval = 12)
#
# hbc2 <- CompareHamBac(100, 400, 1, gamma_shape = 1.5, gamma_mean = 20, ar_coefs = 0.7,
#                       sampling_interval = 1)

hbc3a <- CompareHamBac(100, 400, 1, sim_gamma_shape = 1.5,
                      sim_gamma_mean = 80,
                      ar_coefs = 0.5,
                      sampling_interval = 12,
                      sample_gap = c(201, 349),
                      #K_hamstr = hamstr:::optimal_K(100, 10),
                      K_hamstr = 100,
                      K_bacon = 100,
                      acc_shape = 1.5,
                      mem_mean = 0.5, mem_strength = 10)

hbc3b <- CompareHamBac(100, 400, 1,
                       sim_gamma_shape = 1.5,
                       sim_gamma_mean = 40,
                      ar_coefs = 0.5,
                      sampling_interval = 12,
                      sample_gap = c(201, 349),
                      K_hamstr = hamstr:::optimal_K(100, 10),
                      K_bacon = 100,
                      acc_shape = 1.5,
                      mem_mean = 0.8, mem_strength = 10)

hbc3c <- CompareHamBac(100, 400, 1, gamma_shape = 1.5,
                      gamma_mean = 40,
                      ar_coefs = 0.5,
                      sampling_interval = 12,
                      sample_gap = c(201, 349),
                      K_hamstr = hamstr:::optimal_K(100, 10),
                      K_bacon = 100,
                      acc_shape = 1.5,
                      mem_mean = 0.2, mem_strength = 10)


GetCoverages(hbc3a)
GetCoverages(hbc3b)
GetCoverages(hbc3c)

PlotHamstrBacon(hbc3a)
PlotHamstrBacon(hbc3b)
PlotHamstrBacon(hbc3c)

hbc4 <- CompareHamBac(100, 400, 1, gamma_shape = 1.5,
                      gamma_mean =  c(20, 90), gamma_breaks = 250,
                      ar_coefs = 0.7,
                      sampling_interval = 24,
                      sample_gap = c(225, 300),
                      K_hamstr = hamstr:::optimal_K(100, 10),
                      K_bacon = 100)

hbc5 <- CompareHamBac(100, 400, 1, gamma_shape = 1.5,
                      gamma_mean =  c(90, 20), gamma_breaks = 250,
                      ar_coefs = 0.7,
                      sampling_interval = 24,
                      sample_gap = c(225, 300),
                      K_hamstr = hamstr:::optimal_K(100, 10),
                      K_bacon = 100)

hbc6 <- CompareHamBac(100, 400, 1, gamma_shape = 1.5,
                      gamma_mean =  c(90, 20), gamma_breaks = 250,
                      ar_coefs = 0.7,
                      sampling_interval = 24,
                      sample_gap = c(225, 300),
                      K_hamstr = hamstr:::optimal_K(100, 10),
                      K_bacon = 100,
                      inflate_errors = TRUE)



GetCoverages(hbc3)
GetCoverages(hbc4)
GetCoverages(hbc5)
GetCoverages(hbc6)


PlotHamstrBacon(hbc3)
PlotHamstrBacon(hbc4)
PlotHamstrBacon(hbc5)
PlotHamstrBacon(hbc6)

CompareMedAgeMod(hbc4)

plot(hbc6$hamstr)

comps <- list(hbc1, hbc2, hbc3, hbc4, hbc5, hbc6)

covr <- plyr::ldply(comps, GetCoverages, .progress = TRUE)

covr %>%
  gather(percentile, coverage, -RMSE, -method) %>%
  mutate(percentile_num = readr::parse_number(percentile)) %>%
  ggplot(aes(x = percentile, y = coverage, colour = method)) +
  geom_point(position = position_jitter(width = 0.2)) +
  geom_hline(aes(yintercept = percentile_num/100)) +
  facet_wrap(~percentile, scales = "free_y")

if (parallel::detectCores() >= 4) options(mc.cores = 4)
hbc_list <- replicate(64, CompareHamBac(100, 700, 1,
                                       gamma_shape = 1.5,
                                       gamma_mean = c(rep(10, 200), rep(50, 200), rep(10, 200)),
                                       ar_coefs = 0.7,
                      sampling_interval = 48, K_hamstr = c(10, 10), K_bacon = 600/5),
                      simplify = FALSE)

PlotHamstrBacon(hbc_list[[1]])
GetCoverages(hbc_list[[1]])

names(hbc_list) <- 1:length(hbc_list)

#saveRDS(hbc_list, file = gsub(" ", "_", (paste0("package-dev-scripts/cached-runs/ar0.5_gs3_gm10_50_10_si48_", Sys.time(),".RDS"))))
# Big

lapply(hbc_list[[1]]$bacon, function(x) format(object.size(x), units = "Mb"))

covr <- plyr::ldply(hbc_list, GetCoverages, .progress = "text")

covr %>%
  gather(percentile, coverage, -RMSE, -method, -.id) %>%
  mutate(percentile_num = readr::parse_number(percentile)) %>%
  ggplot(aes(x = percentile, y = coverage, colour = method)) +
  geom_point(position = position_jitter(width = 0.2)) +
  geom_hline(aes(yintercept = percentile_num/100)) #+
  #facet_wrap(~percentile, scales = "free_y")


covr %>%
  gather(percentile, coverage, -RMSE, -method, -.id) %>%
  mutate(percentile_num = readr::parse_number(percentile),
         method_num = as.numeric(as.factor(method))) %>%
  tbl_df() %>%

  ggplot(aes(x = method_num, y = coverage, group = .id)) +
  geom_point() +
  geom_line(alpha = 0.5) +
  geom_hline(aes(yintercept = percentile_num/100)) +
  facet_wrap(~percentile, scales = "free_y") +
  scale_x_continuous(breaks = c(1, 2), labels = c("bacon", "hamstr"))


covr %>%
  gather(percentile, coverage, -RMSE, -method, -.id) %>%
  mutate(percentile_num = readr::parse_number(percentile)) %>%
  ggplot(aes(x = percentile, y = coverage, fill = method)) +
  geom_violin() +
  #geom_point(position = position_jitter(width = 0.2)) +
  geom_hline(aes(yintercept = percentile_num/100)) +
  facet_wrap(~percentile, scales = "free")


covr %>%
  ggplot(aes(x = method, y = RMSE)) +
  geom_boxplot() +
  geom_point(position = position_jitter(width = 0.2))


covr %>%
  mutate(method_num = as.numeric(as.factor(method))) %>%
  ggplot(aes(x = method_num, y = RMSE, group = .id)) +
  geom_point() +
  geom_line() +
  scale_x_continuous(breaks = c(1, 2), labels = c("bacon", "hamstr"))




covr %>%
  group_by(method) %>%
  summarise(across(everything(), mean))



get_bacon_memory <- function(hamstr_bacon_fit){

  b1 <- hamstr_bacon_fit
  b1$posterior <- b1$info$output
  w <- b1$posterior[,ncol(b1$posterior) -1]
  thick <- b1$info$thick
  R <- w^(1/thick)

  dplyr::tibble(iter = 1:nrow(b1$posterior),
                R = R, w = w)

}

get_bacon_memory(hbc3$bacon) %>%
  pivot_longer(cols = c("R", "w")) %>%
  ggplot(aes(x = value, fill = name)) +
    geom_density() +
  scale_x_continuous(limits = c(0, 1))


plot(hbc4$hamstr)


CompareMedAgeMod <- function(sim){

  hamstr = join_mod_dat(sim$hamstr, sim$sim_core)
  bacon = join_mod_dat(sim$bacon, sim$sim_core)

  hb <- bind_rows(hamstr=hamstr, bacon=bacon, .id = "method") %>%
    mutate(error = `50%` - age)

  p <- hb %>%
    ggplot(aes(x = depth, y = `50%` - age, colour = method)) +
    geom_line() +
    geom_hline(yintercept = 0)

  return(list(data = hb, plot = p))

}

err <- CompareMedAgeMod(hbc4)

err$plot

err$data %>%
  ggplot(aes(x = depth, y = abs(`50%` - age), colour = method)) +
  geom_line() +
  geom_hline(yintercept = 0)

err$data %>%
  ggplot(aes(x = (`50%` - age), fill = method)) +
  geom_histogram() +
  geom_vline(xintercept = 0) +
  facet_wrap(~method, ncol = 1)


err$data %>%
  filter(complete.cases(mean)) %>%
  group_by(method) %>%
  summarise(n = n())





PlotHamList <- function(mod.list){

  obs_ages <- mod.list[[1]]$data[c("depth", "obs_age", "obs_err")]

  obs_ages <- dplyr::as_tibble(obs_ages)

  obs_ages <- obs_ages %>%
    dplyr::rename(.,
                  age = obs_age,
                  err  = obs_err) %>%
    dplyr::mutate(age_upr = age + 2*err,
                  age_lwr = age - 2*err)


  s <- lapply(mod.list, function(x) {
    summary(x)
  })

  s <- bind_rows(s, .id = "model")

  alpha.lvl <- 0.5


  p.age.sum <- s %>%
    ggplot2::ggplot(ggplot2::aes(x = depth, y = mean)) +
    ggplot2::geom_ribbon(ggplot2::aes(ymax = `2.5%`, ymin = `97.5%`, fill = model), alpha = alpha.lvl/1.5) +
    ggplot2::geom_ribbon(ggplot2::aes(ymax = `75%`, ymin = `25%`, fill = model), alpha = alpha.lvl) +
    ggplot2::geom_line(ggplot2::aes(y = `50%`, colour = model), lwd = 1) +
    ggplot2::labs(x = "Depth", y = "Age") +
    theme_bw()


  p.age.sum <- p.age.sum +
    ggplot2::geom_linerange(data = obs_ages,
                            ggplot2::aes(x = depth,
                                         ymax = age_upr, ymin = age_lwr), inherit.aes = FALSE,
                            colour = "black") +
    ggplot2::geom_point(data = obs_ages, ggplot2::aes(y = age),
                        colour = "black")

  p.age.sum

}

PlotHamList(tail(hbc4, -1)) +
  geom_line(data = hbc4$sim_core,
            aes(x = depth, y = age), colour = "black")

PlotHamList(list("hamstr_10_10" = hamstr_10_10, hamstr_50 = hamstr_50, "hambac_5" = hambac_5)) #+
# facet_wrap(~model)
